pkg/kubelet/apis/config/validation/validation.go

```diff
@@ -237,10 +235,6 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	}
 	allErrors = append(allErrors, metrics.ValidateShowHiddenMetricsVersion(kc.ShowHiddenMetricsForVersion)...)
 
-	if errs := logs.ValidateLoggingConfiguration(&kc.Logging, field.NewPath("logging")); len(errs) > 0 {
```

Comment:
why don't we want to validate logging configuration here any more?

Reply:
It's done already earlier in `ValidateAndApply` which all binaries should call as soon as possible to configure logging for the other validate calls - several of those have logging calls. Combining those two actions into one was a conscious decision to ensure that the package gets used as intended.

I was a bit torn whether exposing just `Validate` as API call still makes sense. It might be useful in unit tests. In the end I decided to keep the `logs` API as simple as possible, but that's up for debate.

Reply:
I now have a problem for which I do not have a solution yet:
- `pkg/kubelet/apis/config/validation/validation.go` uses `utilfeature.DefaultFeatureGate` when calling `LoggingConfiguration.Validate`
- the new feature gates must be added to  `utilfeature.DefaultMutableFeatureGate` in `validation_test.go`, otherwise looking up the feature panics
- `hack/verify-test-featuregates.sh` [complains about that](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/105797/pull-kubernetes-verify/1520003351708700672):


```
{Script Error ScriptError Test files may not access mutable global feature gates directly:
./pkg/kubelet/apis/config/validation/validation_test.go:79:	logsapi.AddFeatureGates(utilfeature.DefaultMutableFeatureGate)

Use this invocation instead:
  defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()}
```

Do I have to modify `hack/verify-test-featuregates.sh`, for example to allow usage of `utilfeature.DefaultMutableFeatureGate` in `AddFeatureGates` calls?

Reply:
rather than referring to `utilfeature.DefaultFeatureGate.DeepCopy()` inside `ValidateKubeletConfiguration` and therefore having to mess with it in test setup, I'd probably:
1. pass it as an arg to ValidateKubeletConfiguration
2. pass in `utilfeature.DefaultFeatureGate` from non-test call sites
3. construct a test-only instance in validation_test.go and register log gates into that, e.g.:
   ```go
   testFeatureGates := utilfeature.DefaultFeatureGate.DeepCopy()
   logsapi.AddFeatureGates(testFeatureGates)
   ValidateKubeletConfiguration(..., testFeatureGates)
   ```

Reply:
That works, thanks.